openmdao/components/subproblem.py

```diff
@@ -0,0 +1,316 @@
+        to_abs_uname = self._sysdata.to_abs_uname = {}
+        to_abs_pnames = self._sysdata.to_abs_pnames = OrderedDict()
+        to_prom_uname = self._sysdata.to_prom_uname = OrderedDict()
+        to_prom_pname = self._sysdata.to_prom_pname = OrderedDict()
+
+        # Our subproblem has been completely set up. We now just pull
+        # variable metadata from our subproblem
+        subparams = self._problem.root.params
+        subunknowns = self._problem.root.unknowns
+
+        # keep track of params that are actually unknowns in the subproblem
+        self._unknowns_as_params = []
+
+        self._params_dict = self._init_params_dict = OrderedDict()
+        for name in self._prob_params:
+            pathname = self._get_var_pathname(name)
+            if name in subparams:
+                meta = subparams._dat[name].meta
+            elif name in self._problem._dangling:
+                meta = self._rec_get_param_meta(name)
+            else:
+                meta = subunknowns._dat[name].meta
+                self._unknowns_as_params.append(name)
+
+            meta = meta.copy() # don't mess with subproblem's metadata!
+
+            self._params_dict[pathname] = meta
+            meta['pathname'] = pathname
+            del meta['top_promoted_name']
+            to_prom_pname[pathname] = name
+            to_abs_pnames[name] = (pathname,)
+
+        self._unknowns_dict = self._init_unknowns_dict = OrderedDict()
+
+        # if we have params that are really unknowns in the subproblem, we
+        # also add them as unknowns so we can take derivatives
+        for name in self._prob_unknowns:
+            pathname = self._get_var_pathname(name)
+            meta = subunknowns._dat[name].meta.copy()
+            self._unknowns_dict[pathname] = meta
+            meta['pathname'] = pathname
+            del meta['top_promoted_name']
+            to_prom_uname[pathname] = name
+            to_abs_uname[name] = pathname
+
+        to_prom_name.update(to_prom_uname)
+        to_prom_name.update(to_prom_pname)
+
+        self._post_setup_vars = True
+
+        self._sysdata._params_dict = self._params_dict
+        self._sysdata._unknowns_dict = self._unknowns_dict
+
+        return self._params_dict, self._unknowns_dict
+
+    def solve_nonlinear(self, params, unknowns, resids):
+        """Sets params into the sub-problem, runs the
+        sub-problem, and updates our unknowns with values
+        from the sub-problem.
+
+        Args
+        ----
+        params : `VecWrapper`, optional
+            `VecWrapper` containing parameters. (p)
+
+        unknowns : `VecWrapper`, optional
+            `VecWrapper` containing outputs and states. (u)
+
+        resids : `VecWrapper`, optional
+            `VecWrapper` containing residuals. (r)
+        """
+
+        try:
+            # set params into the subproblem
+            prob = self._problem
+            for name in self._prob_params:
+                prob[name] = params[name]
```

Comment:
I didn't see the need here for a user-centric error message, because I've already checked that the contents of self._prob_params are all variables that exist and are allowed to be set in the subproblem, and 'params' is always (except possibly in our tests) the same as self.params, which is built using the same list of variables found in self._prob_params, so it seems that if a KeyError occurred here it would be due to some framework error or possibly an error in one of our tests, but wouldn't be a user error because users don't call solve_nonlinear directly.